File: src/main/java/io/scriptor/fdt/Prop.java
package io.scriptor.fdt;

import org.jetbrains.annotations.NotNull;

import java.nio.ByteBuffer;

import static io.scriptor.fdt.Constant.FDT_PROP;

public record Prop(@NotNull String name, @NotNull Writable data) {

    public void write(final @NotNull ByteBuffer buffer, final @NotNull StringTable strings) {
        final var nameoff = strings.push(name);

        buffer.putInt(FDT_PROP)
              .putInt(data.size())
              .putInt(nameoff);

        data.write(buffer);
        while ((buffer.position() & 3) != 0) {
            buffer.put((byte) 0);
        }
    }

    public int size() {
        int size = 0;
        size += Integer.SIZE;
        size += Integer.SIZE;
        size += Integer.SIZE;
        size += data.size();
        size = (size + 3) & ~3;
        return size;
    }
}

File: src/main/java/io/scriptor/fdt/Prop.kt
package io.scriptor.fdt

import java.nio.ByteBuffer

data class Prop(val name: String, val data: Writable) {
    fun write(buffer: ByteBuffer, strings: StringTable) {
        val nameoff = strings.push(name)

        buffer
            .putInt(Constant.FDT_PROP.toInt())
            .putInt(data.size())
            .putInt(nameoff)

        data.write(buffer)
        while ((buffer.position() and 3) != 0) {
            buffer.put(0.toByte())
        }
    }

    fun size(): Int {
        var size = 0
        size += Integer.SIZE
        size += Integer.SIZE
        size += Integer.SIZE
        size += data.size()
        size = (size + 3) and 3.inv()
        return size
    }
}
